packages/react-reconciler/src/ReactFiberTreeReflection.js

```diff
@@ -353,17 +353,137 @@ function traverseFragmentInstanceChildren<A, B, C>(
   }
 }
 
-export function getFragmentParentHostInstance(fiber: Fiber): null | Instance {
+export function groupFragmentChildrenByScrollContainer(
+  fragmentFiber: Fiber,
+  detectNewHostScrollContainer: (fiber: Fiber) => boolean,
+): Array<Array<Fiber>> {
+  const groups: Array<Array<Fiber>> = [[]];
+  groupFragmentChildrenByScrollContainer_recursive(
+    fragmentFiber.child,
+    groups,
+    detectNewHostScrollContainer,
+  );
+  return groups;
+}
+
+function groupFragmentChildrenByScrollContainer_recursive(
+  child: Fiber | null,
+  groups: Array<Array<Fiber>>,
+  detectNewHostScrollContainer: (fiber: Fiber) => boolean,
+): void {
+  let withinPortal: boolean = false;
+  while (child !== null) {
+    if (child.tag === HostComponent) {
+      const needsNewScrollContainer = detectNewHostScrollContainer(child);
+      if (needsNewScrollContainer) {
+        if (groups[groups.length - 1].length !== 0) {
+          groups.push([child]);
+        } else {
+          groups[groups.length - 1].push(child);
+        }
+        groups.push([]);
+      } else {
+        groups[groups.length - 1].push(child);
+      }
+    } else if (
+      child.tag === OffscreenComponent &&
+      child.memoizedState !== null
+    ) {
+      // Skip hidden subtrees
+    } else {
+      // Add children of a portal into their own group
```

Comment:
There's not really anything special about portals in that you can have one portal point to the same parent as the regular children and a different portal to document.body. Really what matters is the `.parentNode` of the DOM node inside the portal.